pallets/common/src/protocol_fee.rs

```diff
@@ -0,0 +1,34 @@
+use codec::{Decode, Encode};
+use frame_support::dispatch::DispatchResult;
+use polymesh_primitives::Signatory;
+#[cfg(feature = "std")]
+use sp_runtime::{Deserialize, Serialize};
+
+/// Protocol fee operations.
+#[derive(Decode, Encode, Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub enum ProtocolOp {
+    AssetRegisterTicker,
+    AssetIssue,
+    AssetAddDocument,
+    AssetCreateToken,
+    DividendNew,
+    GeneralTmAddActiveRule,
+    IdentityRegisterDid,
+    IdentityCddRegisterDid,
+    IdentityAddClaim,
+    IdentitySetMasterKey,
+    IdentityAddSigningItem,
+    MipsPropose,
+    VotingAddBallot,
+}
+
+/// Common interface to protocol fees for runtime modules.
+pub trait ChargeProtocolFee<AccountId> {
+    /// Computes the fee of the operation and charges it to the given signatory.
+    fn charge_fee(signatory: &Signatory, op: ProtocolOp) -> DispatchResult;
```

Comment:
In most places where this is being called, the parent function can revert after calling this. We don't want that to happen. Perhaps we should add a function called `can_charge_fee` that makes sure the user has enough balance to pay the fee and we call it at the start of the function (This can be a signedextension but that'd make it more complex so I'd avoid it in this PR). We can then call `charge_fee` at the end of the parent function so that fee is charged only when there is no chance of the parent function throwing an error.

Reply:
You are right that it can be another PR. I think adding an extra check would duplicate the existing functionality of `charge_fee` which already returns errors `InsufficientIdentityBalance` and `InsufficientAccountBalance`, as far as existing calls are concerned. In all the existing calls, `charge_fee` comes after all other checks and before any mutable accesses.

Reply:
In general, `can_charge_fee` isn't a solution since the balance can decrease after `can_charge_fee` returns `true` but before `charge_fee`. So that `charge_fee` can still fail despite the success of `can_charge_fee`.